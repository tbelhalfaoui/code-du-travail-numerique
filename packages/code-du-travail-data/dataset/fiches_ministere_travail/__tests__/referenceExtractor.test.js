const fs = require("fs");
const path = require("path");
const { classifyTokens, extractReferences } = require("../referenceExtractor");
const { resolveReferences } = require("../referenceResolver");

const annotatedTokens = fs
  .readFileSync(path.join(__dirname, "referenceExtractor.test.txt"))
  .toString()
  .split("\n");

const tokens = [];
const labels = [];

annotatedTokens.map(line => {
  const [t, l] = line.split("\t");
  tokens.push(t);
  labels.push(l);
});

const testCases = [
  {
    input:
      "les modalités fixées par les articles L. 2313-8 et R. 2313-3 à R. 2313-6 du code du travail ainsi que le L. 1251-18",
    expected: ["L. 2313-8", "R. 2313-3 à R. 2313-6"]
  },
  {
    input:
      "l’allocation de remplacement pour maternité ou paternité, prévues aux articles L. 613-19 à L.613-19-2 et L. 722-8 à 25 du code de la sécurité sociale, aux articles L. 732-10 à L. 732-12-1 du code rural et à l’article 17 de la loi n° 97-1051 du 18 novembre 1997 d’orientation sur la pêche maritime et les cultures marines",
    expected: [
      "L. 613-19 à L.613-19-2",
      "L. 722-8 à 25",
      "L. 732-10 à L. 732-12-1"
    ]
  },
  {
    input: `Article D212 du code penal et article R413 du code civil`,
    expected: ["D212", "R413"]
  },
  { input: `Article D212`, expected: ["D212"] },
  { input: `Article D-212`, expected: ["D-212"] },
  { input: `Article D.212`, expected: ["D.212"] },
  { input: `Article D212-3`, expected: ["D212-3"] },
  { input: `Article D-212-4`, expected: ["D-212-4"] },
  { input: `Article X*212-4`, expected: [] },
  { input: `Article D. 212-4`, expected: ["D. 212-4"] },
  { input: `Article D.  212-4`, expected: ["D. 212-4"] },
  { input: `Article D.212-5`, expected: ["D.212-5"] },
  { input: `Article D.212-5-6`, expected: ["D.212-5-6"] },
  { input: `Article D.212-5-6-7`, expected: ["D.212-5-6-7"] },
  { input: `Article XD212`, expected: [] }
];

test("should extract article tokens in examples", () => {
  expect(
    testCases.map(({ input }) => extractReferences(input))
  ).toMatchSnapshot();
});

test("should success with actual real life set", () => {
  const predictions = classifyTokens(tokens);
  // keep this to easily check prediction errors :
  /*
  const report = tokens.map((t, i) => {
    const label = labels[i];
    const prediction = predictions[i];
    const ok = prediction == label;
    return [
      `${ok ? "\x1b[0m" : "\x1b[31m"}`,
      `${t}\t${label}\t${prediction}\n`
    ];
  });
  console.log(...report.flat());
  */
  expect(predictions).toEqual(labels);
});

it("should find with code for actual real life set", () =>
  expect(extractReferences(tokens.join(" "))).toMatchSnapshot());

it("should resolve example codes", () => {
  const refs = extractReferences(testCases[0].input);
  expect(resolveReferences(refs)).toMatchSnapshot();
});
