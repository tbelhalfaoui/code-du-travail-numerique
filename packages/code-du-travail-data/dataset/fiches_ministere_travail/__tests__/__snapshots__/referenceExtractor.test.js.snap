// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should extract article tokens in examples 1`] = `
Array [
  Array [
    Object {
      "article": "L. 2313-8",
      "code": "code du travail",
    },
    Object {
      "article": "R. 2313-3 à R. 2313-6",
      "code": "code du travail",
    },
  ],
  Array [
    Object {
      "article": "L. 613-19 à L.613-19-2",
      "code": undefined,
    },
    Object {
      "article": "L. 722-8 à 25",
      "code": undefined,
    },
    Object {
      "article": "L. 732-10 à L. 732-12-1",
      "code": undefined,
    },
  ],
  Array [
    Object {
      "article": "D212",
      "code": undefined,
    },
    Object {
      "article": "R413",
      "code": undefined,
    },
  ],
  Array [
    Object {
      "article": "D212",
      "code": undefined,
    },
  ],
  Array [
    Object {
      "article": "D-212",
      "code": undefined,
    },
  ],
  Array [
    Object {
      "article": "D.212",
      "code": undefined,
    },
  ],
  Array [
    Object {
      "article": "D212-3",
      "code": undefined,
    },
  ],
  Array [
    Object {
      "article": "D-212-4",
      "code": undefined,
    },
  ],
  Array [],
  Array [
    Object {
      "article": "D. 212-4",
      "code": undefined,
    },
  ],
  Array [
    Object {
      "article": "D. 212-4",
      "code": undefined,
    },
  ],
  Array [
    Object {
      "article": "D.212-5",
      "code": undefined,
    },
  ],
  Array [
    Object {
      "article": "D.212-5-6",
      "code": undefined,
    },
  ],
  Array [
    Object {
      "article": "D.212-5-6-7",
      "code": undefined,
    },
  ],
  Array [],
]
`;

exports[`should find with code for actual real life set 1`] = `
Array [
  Object {
    "article": "L. 2313-8",
    "code": "code du travail",
  },
  Object {
    "article": "R. 2313-3 à R. 2313-6",
    "code": "code du travail",
  },
  Object {
    "article": "R. 2242-7",
    "code": "code du travail",
  },
  Object {
    "article": "L. 2242-8",
    "code": "code du travail",
  },
  Object {
    "article": "R. 2242-2 à R. 2242-11",
    "code": undefined,
  },
  Object {
    "article": "L. 2242-9",
    "code": "code du travail",
  },
  Object {
    "article": "R. 2242-9 à R. 2242-11",
    "code": "code du travail",
  },
  Object {
    "article": "L. 135-1",
    "code": "code de la sécurité sociale",
  },
  Object {
    "article": "D. 1142-2",
    "code": "code du travail",
  },
  Object {
    "article": "D. 1142-2-1",
    "code": "code du travail",
  },
  Object {
    "article": "L. 4521-1",
    "code": "code du travail",
  },
  Object {
    "article": "D. 1142-2",
    "code": "code du travail",
  },
  Object {
    "article": "D. 1142-2-1",
    "code": "code du travail",
  },
  Object {
    "article": "R. 1234-9",
    "code": "code du travail",
  },
  Object {
    "article": "L. 130-1",
    "code": "code de la sécurité sociale",
  },
  Object {
    "article": "R. 130-1",
    "code": "code de la sécurité sociale",
  },
  Object {
    "article": "D. 8223-2",
    "code": "code du travail",
  },
  Object {
    "article": "R. 1263-3",
    "code": "code du travail",
  },
  Object {
    "article": "R. 1263-4",
    "code": "code du travail",
  },
  Object {
    "article": "R. 1263-6",
    "code": "code du travail",
  },
  Object {
    "article": "L. 613-19 à L. 613-19-2",
    "code": "code de la sécurité sociale",
  },
  Object {
    "article": "L. 722-8 à L. 722-8-3",
    "code": "code de la sécurité sociale",
  },
  Object {
    "article": "L. 732-10 à L. 732-12-1",
    "code": undefined,
  },
  Object {
    "article": "L. 4412-2",
    "code": undefined,
  },
  Object {
    "article": "R. 4412-97/II",
    "code": undefined,
  },
  Object {
    "article": "L. 2315-24",
    "code": "code du travail",
  },
  Object {
    "article": "L. 2315-38 à 40",
    "code": undefined,
  },
  Object {
    "article": "L1111-2",
    "code": undefined,
  },
  Object {
    "article": "L. 2314-6",
    "code": "code du travail",
  },
  Object {
    "article": "R. 2314-6",
    "code": "code du travail",
  },
  Object {
    "article": "L 2312-36",
    "code": "code du travail",
  },
  Object {
    "article": "R. 2312-8",
    "code": undefined,
  },
  Object {
    "article": "R. 2312-10 à R. 2312-14",
    "code": undefined,
  },
  Object {
    "article": "R. 2312-9",
    "code": undefined,
  },
  Object {
    "article": "R. 2312-10 à R. 2312-14",
    "code": undefined,
  },
]
`;
